code/datums/extensions/health/_health.dm

```diff
@@ -0,0 +1,175 @@
+ */
+/datum/extension/health
+	base_type = /datum/extension/health
+	expected_type = /atom
+
+	/// Current health, from `0` to `max_health`.
+	var/current_health
+
+	/// Maximum health. `0` or `NULL` means no health handling.
+	var/max_health
+
+	/// Whether or not the death state has been set.
+	var/death_state = FALSE
+
+	/// LAZY List of damage type resistance or weakness multipliers, decimal form. Only applied to health reduction.
+	var/list/damage_resistances
+
+	/// Minimum damage required to actually affect health in `can_damage_health()`
+	var/min_damage_threshhold = 0
+
+
+/**
+ * Retrieves the extension's `current_health`.
+ * Provided as a proc for override use.
+ */
+/datum/extension/health/proc/current_health()
+	return current_health
+
+
+/**
+ * Retrieves the extension's `max_health`.
+ * Provided as a proc for override use.
+ */
+/datum/extension/health/proc/max_health(real = FALSE)
+	var/atom/A = holder
+	return real ? max_health : A.get_max_health()
+
+
+/**
+ * Retrieves the extension's damage (Difference between `max_health` and `current_health`).
+ * Provided as a proc for override use.
+ */
+/datum/extension/health/proc/current_damage(damage_type = null)
+	return max_health - current_health
+
+
+/**
+ * Retrieves the extension's death state.
+ * Provided as a proc for override use.
+ */
+/datum/extension/health/proc/get_death_state()
+	return death_state
+
+
+/**
+ * Sets a new death_state.
+ * Returns `TRUE` of the new state is different from the old state, `FALSE` otherwise.
+ * Also calls `handle_death_change()` on the holder.
+ */
+/datum/extension/health/proc/set_death_state(new_death_state)
+	if (death_state == new_death_state)
+		return FALSE
+	death_state = new_death_state
+	var/atom/A = holder
+	A.handle_death_change(new_death_state)
+	return TRUE
+
+
+/**
+ * Checks the current condition of health and sets the death state accordingly.
+ * Returns `TRUE` if `death_state` was changed, `FALSE` otherwise.
+ */
+/datum/extension/health/proc/update_death_state()
+	if (current_health() == 0)
+		return set_death_state(TRUE)
+	else
+		return set_death_state(FALSE)
+
+
+/**
+ * Modifies `current_health` by the given modifier.
+ * Returns `TRUE` if `death_state` was changed, `FALSE` otherwise.
+ */
+/datum/extension/health/proc/mod_health(health_mod, damage_type = null)
+	SHOULD_CALL_PARENT(TRUE)
+
+	// Apply resistances to negative modifiers
+	if (health_mod < 0 && damage_type && LAZYISIN(damage_resistances, damage_type))
+		health_mod = health_mod * damage_resistances[damage_type]
+
+	health_mod = round(health_mod, 1)
+	if (health_mod == 0)
+		return FALSE // No health change, no further processing needed
+	current_health = Clamp(current_health + health_mod, 0, max_health())
+
+	var/atom/A = holder
+	if (health_mod > 0)
+		A.post_restore_health(health_mod, damage_type)
+	else if (health_mod < 0)
+		A.post_damage_health(abs(health_mod), damage_type)
+	A.post_health_change(health_mod, damage_type)
```

Comment:
This block highlights a concern I have about the current implementation, in that it's adding tons of proc calls for stuff that isn't actually implemented. You could just override this proc and handle post-health logic there, or on the extension, rather than making every atom call at least two additional procs for no reason.

Reply:
`post_health_change()` was added because an atom may want to do additional things after the extensions modify its health, without needing to create additional subtypes of the health extension.

The other two `post_*_health()` procs were mostly just a convenience thing but I can remove those.

Reply:
Additional subtypes of the health extension would be preferable to future-proofing that causes proc overhead, imo.